all: some resource leaks detected due to missing .Close() calls

### What is the issue?

I ran Orijtech Inc's resource leak static analyzer [Static Major](https://github.com/marketplace/actions/staticmajor-analyzer) on Tailscale and it produced this report

```shell
$ staticmajor -resleak=true -resleak.fuzzy ./...
/github.com/tailscale/tailscale/chirp/chirp.go:41:3: leaking resource created on line 0
/github.com/tailscale/tailscale/client/tailscale/example/servetls/servetls.go:29:1: leaking resource created on line 0
/github.com/tailscale/tailscale/cmd/derper/derper.go:316:4: leaking resource created on line 0
/github.com/tailscale/tailscale/cmd/hello/hello.go:99:3: leaking resource created on line 0
/github.com/tailscale/tailscale/wgengine/monitor/monitor.go:96:3: leaking resource created on line 0
/github.com/tailscale/tailscale/control/controlclient/auto.go:97:3: leaking resource created on line 92
/github.com/tailscale/tailscale/control/controlclient/noise.go:103:3: leaking resource created on line 0
/github.com/tailscale/tailscale/portlist/poller.go:50:3: leaking resource created on line 0
/github.com/tailscale/tailscale/net/ping/ping.go:67:4: leaking resource created on line 57
/github.com/tailscale/tailscale/net/netcheck/netcheck.go:1193:3: leaking resource created on line 1190
/github.com/tailscale/tailscale/net/netcheck/netcheck.go:1202:3: leaking resource created on line 1190
/github.com/tailscale/tailscale/net/netcheck/netcheck.go:1225:3: leaking resource created on line 1190
/github.com/tailscale/tailscale/net/netcheck/netcheck.go:1230:3: leaking resource created on line 1190
/github.com/tailscale/tailscale/net/netcheck/netcheck.go:1238:3: leaking resource created on line 1190
/github.com/tailscale/tailscale/net/netcheck/netcheck.go:1243:3: leaking resource created on line 1190
/github.com/tailscale/tailscale/net/netcheck/netcheck.go:1249:2: leaking resource created on line 1190
/github.com/tailscale/tailscale/ipn/ipnlocal/local.go:3626:1: leaking resource created on line 0
/github.com/tailscale/tailscale/wgengine/netstack/netstack.go:943:4: leaking resource created on line 926
/github.com/tailscale/tailscale/wgengine/netstack/netstack.go:937:3: leaking resource created on line 926
/github.com/tailscale/tailscale/wgengine/netstack/netstack.go:933:3: leaking resource created on line 926
/github.com/tailscale/tailscale/tsnet/tsnet.go:504:3: leaking resource created on line 0
/github.com/tailscale/tailscale/cmd/tailscale/cli/cert.go:61:3: leaking resource created on line 0
/github.com/tailscale/tailscale/cmd/tailscale/cli/file.go:365:4: leaking resource created on line 346
/github.com/tailscale/tailscale/cmd/tailscale/cli/file.go:368:4: leaking resource created on line 346
/github.com/tailscale/tailscale/cmd/tailscale/cli/file.go:370:3: leaking resource created on line 346
/github.com/tailscale/tailscale/cmd/tailscale/cli/file.go:379:5: leaking resource created on line 346
/github.com/tailscale/tailscale/cmd/tailscale/cli/file.go:354:3: leaking resource created on line 346
/github.com/tailscale/tailscale/cmd/tailscale/cli/web.go:136:3: leaking resource created on line 0
/github.com/tailscale/tailscale/ipn/ipnserver/server.go:316:3: leaking resource created on line 0
/github.com/tailscale/tailscale/tempfork/gliderlabs/ssh/ssh.go:117:4: leaking resource created on line 0
/github.com/tailscale/tailscale/tempfork/gliderlabs/ssh/ssh.go:120:2: leaking resource created on line 0
/github.com/tailscale/tailscale/tempfork/gliderlabs/ssh/ssh.go:130:4: leaking resource created on line 0
/github.com/tailscale/tailscale/tempfork/gliderlabs/ssh/ssh.go:133:2: leaking resource created on line 0
/github.com/tailscale/tailscale/ssh/tailssh/tailssh.go:1302:3: leaking resource created on line 0
/github.com/tailscale/tailscale/ssh/tailssh/tailssh.go:1306:3: leaking resource created on line 0
/github.com/tailscale/tailscale/ssh/tailssh/tailssh.go:1310:3: leaking resource created on line 0
/github.com/tailscale/tailscale/ssh/tailssh/tailssh.go:1341:3: leaking resource created on line 0
/github.com/tailscale/tailscale/ssh/tailssh/tailssh.go:1347:3: leaking resource created on line 0
/github.com/tailscale/tailscale/cmd/tailscaled/debug.go:212:3: leaking resource created on line 202
/github.com/tailscale/tailscale/cmd/tailscaled/debug.go:212:3: leaking resource created on line 203
/github.com/tailscale/tailscale/cmd/tailscaled/debug.go:219:3: leaking resource created on line 202
/github.com/tailscale/tailscale/cmd/tailscaled/debug.go:219:3: leaking resource created on line 203
/github.com/tailscale/tailscale/cmd/tailscaled/debug.go:222:2: leaking resource created on line 202
/github.com/tailscale/tailscale/cmd/tailscaled/debug.go:222:2: leaking resource created on line 203
/github.com/tailscale/tailscale/cmd/tailscaled/debug.go:89:3: leaking resource created on line 87
/github.com/tailscale/tailscale/cmd/tailscaled/debug.go:104:3: leaking resource created on line 87
/github.com/tailscale/tailscale/cmd/tailscaled/tailscaled.go:568:4: leaking resource created on line 561
/github.com/tailscale/tailscale/cmd/tailscaled/tailscaled.go:608:1: leaking resource created on line 0
/github.com/tailscale/tailscale/tsnet/example/tshello/tshello.go:49:1: leaking resource created on line 0
/github.com/tailscale/tailscale/wgengine/bench/bench.go:133:1: leaking resource created on line 0
```

Some of those reports are false positives but shall eventually we shall make updates, but some of them are real resource leaks

### Steps to reproduce

Run staticmajor as a Github action per https://github.com/marketplace/actions/staticmajor-analyzer

### Are there any recent changes that introduced the issue?

Not really

### OS

_No response_

### OS version

_No response_

### Tailscale version

_No response_

### Bug report

_No response_

--- a/chirp/chirp.go
+++ b/chirp/chirp.go
@@ -24,11 +24,17 @@ func New(socket string) (*BIRDClient, error) {
 	return newWithTimeout(socket, responseTimeout)
 }
 
-func newWithTimeout(socket string, timeout time.Duration) (*BIRDClient, error) {
+func newWithTimeout(socket string, timeout time.Duration) (_ *BIRDClient, err error) {
 	conn, err := net.Dial("unix", socket)
 	if err != nil {
 		return nil, fmt.Errorf("failed to connect to BIRD: %w", err)
 	}
+	defer func() {
+		if err != nil {
+			conn.Close()
+		}
+	}()
+
 	b := &BIRDClient{
 		socket:  socket,
 		conn:    conn,

--- a/cmd/tailscale/cli/configure-host.go
+++ b/cmd/tailscale/cli/configure-host.go
@@ -48,7 +48,7 @@ func runConfigureHost(ctx context.Context, args []string) error {
 	if uid := os.Getuid(); uid != 0 {
 		return fmt.Errorf("must be run as root, not %q (%v)", os.Getenv("USER"), uid)
 	}
-	hi:= hostinfo.New()
+	hi := hostinfo.New()
 	isDSM6 := strings.HasPrefix(hi.DistroVersion, "6.")
 	isDSM7 := strings.HasPrefix(hi.DistroVersion, "7.")
 	if !isDSM6 && !isDSM7 {

--- a/net/netcheck/netcheck.go
+++ b/net/netcheck/netcheck.go
@@ -1188,6 +1188,8 @@ func (c *Client) measureHTTPSLatency(ctx context.Context, reg *tailcfg.DERPRegio
 	var ip netip.Addr
 
 	dc := derphttp.NewNetcheckClient(c.logf)
+	defer dc.Close()
+
 	tlsConn, tcpConn, node, err := dc.DialRegionTLS(ctx, reg)
 	if err != nil {
 		return 0, ip, err

--- a/net/nettest/listener.go
+++ b/net/nettest/listener.go
@@ -61,7 +61,7 @@ func (l *Listener) Accept() (net.Conn, error) {
 // The provided Context must be non-nil. If the context expires before the
 // connection is complete, an error is returned. Once successfully connected
 // any expiration of the context will not affect the connection.
-func (l *Listener) Dial(ctx context.Context, network, addr string) (net.Conn, error) {
+func (l *Listener) Dial(ctx context.Context, network, addr string) (_ net.Conn, err error) {
 	if !strings.HasSuffix(network, "tcp") {
 		return nil, net.UnknownNetworkError(network)
 	}
@@ -72,6 +72,13 @@ func (l *Listener) Dial(ctx context.Context, network, addr string) (net.Conn, er
 		}
 	}
 	c, s := NewConn(addr, bufferSize)
+	defer func() {
+		if err != nil {
+			c.Close()
+			s.Close()
+		}
+	}()
+
 	select {
 	case <-ctx.Done():
 		return nil, ctx.Err()

--- a/wgengine/netstack/netstack.go
+++ b/wgengine/netstack/netstack.go
@@ -987,16 +987,19 @@ func (ns *Impl) acceptUDP(r *udp.ForwarderRequest) {
 	}
 	dstAddr, ok := ipPortOfNetstackAddr(sess.LocalAddress, sess.LocalPort)
 	if !ok {
+		ep.Close()
 		return
 	}
 	srcAddr, ok := ipPortOfNetstackAddr(sess.RemoteAddress, sess.RemotePort)
 	if !ok {
+		ep.Close()
 		return
 	}
 
 	// Handle magicDNS traffic (via UDP) here.
 	if dst := dstAddr.Addr(); dst == magicDNSIP || dst == magicDNSIPv6 {
 		if dstAddr.Port() != 53 {
+			ep.Close()
 			return // Only MagicDNS traffic runs on the service IPs for now.
 		}
 

